### PR TITLE
Align eth1data Majority Vote with the spec

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Algorithm re-implementation

**What does this PR do? Why is it needed?**

Original implementation of the eth1data Voting with the Majority algorithm was based on our design doc. One Discord user pointed out a difference between this implementation and the spec. After looking closer at the spec, it turns out that there were other discrepancies between the spec and the code.

This PR aims to align our implementation of the algorithm with the spec. I added several new tests to make sure all potencial cases are covered.

**Which issues(s) does this PR fix?**

N/A

**Other notes for review**

I needed to modify the POW chain mock because the naive implementation would not work for some test cases.

### CODE REVIEW

**Review by farazdagi — APPROVED** [review]
- Great tests!
- I've limited understanding of underlying algorithm, so it is probably prudent to get +1 update from team member, who is more familiar with it (before merging).

File: beacon-chain/rpc/validator/proposer.go [hunk 1]
```diff
@@ -246,51 +249,66 @@ func (vs *Server) eth1DataMajorityVote(ctx context.Context, beaconState *stateTr
...
 	}
 	if len(inRangeVotes) == 0 {
-		eth1Data, err := vs.defaultEth1DataResponse(ctx, currentPeriodBlockNumber)
-		if err != nil {
-			log.WithError(err).Error("Failed to get eth1 data from current period block number")
-			return vs.randomETH1DataVote(ctx)
+		// If latest block is in range and it doesn't undo deposits, then choose that block.
+		// Otherwise take the current eth1data.
+		currentETH1Data := vs.HeadFetcher.HeadETH1Data()
+		if timeOfLastBlockByLatestValidTime >= earliestValidTime {
```
Comment (nisdas) [inline beacon-chain/rpc/validator/proposer.go:296]: Isn't this already a given ?
```go
	earliestValidTime := votingPeriodStartTime - 2*params.BeaconConfig().SecondsPerETH1Block*uint64(eth1FollowDistance)
	latestValidTime := votingPeriodStartTime - params.BeaconConfig().SecondsPerETH1Block*uint64(eth1FollowDistance)
```
Unless the assumption is `timeOfLastBlockByLatestValidTime` untrusted, then the check should be shifted above when we first retrieve it in `BlockNumberByTimestamp`

**Review by nisdas — APPROVED** [review]

File: beacon-chain/powchain/testing/mock.go [hunk 2]
```diff
@@ -78,7 +90,14 @@ func (m *POWChain) BlockTimeByHeight(_ context.Context, height *big.Int) (uint64
 
 // BlockNumberByTimestamp --
 func (m *POWChain) BlockNumberByTimestamp(_ context.Context, time uint64) (*big.Int, error) {
-	return m.BlockNumberByTime[time], nil
+	var chosenTime uint64
+	var chosenNumber *big.Int
+	for t, num := range m.BlockNumberByTime {
+		if chosenNumber == nil || (t > chosenTime && t <= time) {
```
[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/powchain/testing/mock.go:96]: :eyes:

[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/powchain/testing/mock.go:96]: I actually had another bug here. The check `chosenNumber == nil` caused non-deterministic test results. At the beginning of the loop this condition is always true, but there is no guaranteed order for map retrieval, so a block with incorrect time sometimes got assigned to `chosenNumber`.

**Review by nisdas — APPROVED** [review]

**Pull request merged**